File: core/plugin/__DEV/Python/n_gfUtilVectorToEuler.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfUtilAngularMath node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.

Requirements:
    Maya 2017 or above.

Todo:
    * Convert angle values to radians.

This code supports Pylint. Rc file in project.
"""
import maya.api._OpenMaya_py2 as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class VectorToEuler(om2.MPxNode):
    """ Main class of gfUtilVectorToEuler node. """

    kNodeName = ""
    kNodeClassify = ""
    kNodeID = ""

    inVector = om2.MObject()
    outEuler = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return VectorToEuler()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to VectorToEuler class. Instances of VectorToEuler will use these attributes to create plugs
        for use in the compute() method.
        """
        uAttr = om2.MFnUnitAttribute()
        nAttr = om2.MFnNumericAttribute()

        vectorX = nAttr.create("vectorX", "vecx", om2.MFnNumericData.kDouble, 0.0)
        vectorY = nAttr.create("vectorY", "vecy", om2.MFnNumericData.kDouble, 0.0)
        vectorZ = nAttr.create("vectorZ", "vecz", om2.MFnNumericData.kDouble, 0.0)
        VectorToEuler.inVector = nAttr.create("vector", "vec", vectorX, vectorY, vectorZ)
        INPUT_ATTR(nAttr)

        outEulerX = uAttr.create("outEulerX", "oex", om2.MFnUnitAttribute.kAngle, 0.0)
        outEulerY = uAttr.create("outEulerY", "oey", om2.MFnUnitAttribute.kAngle, 0.0)
        outEulerZ = uAttr.create("outEulerZ", "oez", om2.MFnUnitAttribute.kAngle, 0.0)
        VectorToEuler.outEuler = nAttr.create("outEuler", "oe", outEulerX, outEulerY, outEulerZ)
        OUTPUT_ATTR(nAttr)

        VectorToEuler.addAttribute(VectorToEuler.inVector)
        VectorToEuler.addAttribute(VectorToEuler.outEuler)
        VectorToEuler.attributeAffects(VectorToEuler.inVector, VectorToEuler.outEuler)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use
        if plug != VectorToEuler.outEuler:
            return om2.kUnknownParameter

        vVector = dataBlock.inputValue(VectorToEuler.inVector).asVector()
        eEuler = om2.MEulerRotation(
            om2.MAngle(vVector.x, om2.MAngle.kDegrees).asRadians(),
            om2.MAngle(vVector.y, om2.MAngle.kDegrees).asRadians(),
            om2.MAngle(vVector.z, om2.MAngle.kDegrees).asRadians(),
            om2.MEulerRotation.kXYZ
        )

        outEulerHandle = dataBlock.outputValue(VectorToEuler.outEuler)
        outEulerHandle.setMVector(eEuler.asVector())
        outEulerHandle.setClean()

File: plug-ins/dev/cpp/source/n_gfUtilVectorToEuler.cpp
#include "headers/n_gfUtilVectorToEuler.h"


// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
VectorToEuler::VectorToEuler() {}

// Destructor.
VectorToEuler::~VectorToEuler() {}

MObject VectorToEuler::inVector;
MObject VectorToEuler::outEuler;
MObject VectorToEuler::outEulerX;
MObject VectorToEuler::outEulerY;
MObject VectorToEuler::outEulerZ;


void* VectorToEuler::creator(){
    // Maya creator function.
    return new VectorToEuler();
}

MStatus VectorToEuler::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to TestNode class. Instances of TestNode will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnUnitAttribute uAttr;
    MFnNumericAttribute nAttr;

    inVector = nAttr.createPoint("vector", "vec", &status);
    INPUT_ATTR(nAttr);

    outEulerX = uAttr.create("outEulerX", "oex", MFnUnitAttribute::kAngle, 0.0, &status);
    outEulerY = uAttr.create("outEulerY", "oey", MFnUnitAttribute::kAngle, 0.0, &status);
    outEulerZ = uAttr.create("outEulerZ", "oez", MFnUnitAttribute::kAngle, 0.0, &status);
    outEuler = nAttr.create("outEuler", "oe", outEulerX, outEulerY, outEulerZ, &status);
    OUTPUT_ATTR(nAttr);

    addAttribute(inVector);
    addAttribute(outEuler);
    attributeAffects(inVector, outEuler);

    return status;
}

MStatus VectorToEuler::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    float3 &pos = dataBlock.inputValue(inVector).asFloat3();
    double euler[3];
    for (unsigned int i = 0; i < 3; i++){
        euler[i] = pos[i] * (M_PI / 180.0);
    }

    MDataHandle outEulerHandle = dataBlock.outputValue(outEuler);
    outEulerHandle.set3Double(euler[0], euler[1], euler[2]);
    outEulerHandle.setClean();

    return MStatus::kSuccess;
}